### PR TITLE
First draft of OpenAttestationEmbeddedRenderer

### PR DESCRIPTION
**Pull request by HJunyuan**
***
<a href="https://pr-preview.s3.amazonaws.com/Open-Attestation/vc-render-method/pull/13.html" title="Last updated on Jun 18, 2024, 2:12 AM UTC (65cfc33)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/vc-render-method/13/3fe3997...Open-Attestation:65cfc33.html" title="Last updated on Jun 18, 2024, 2:12 AM UTC (65cfc33)">Diff</a>

### CODE REVIEW

**Review by msporny — CHANGES_REQUESTED** [review]
A group of us discussed this on the VC API call today. 

General feedback was: the PR seems to be good enough to merge into the specification. The API section should probably go into the OAER section as a subsection. After that reformat is done we can merge this PR.

File: index.html [hunk 1]
```diff
@@ -350,6 +350,141 @@ <h4>SvgRenderingTemplate</h4>
...
+      <p>
+        OpenAttestationEmbeddedRenderer is used by an issuer to render a verifiable
+        credential. The verifiable credential is rendered in HTML within an embedded
+        <code>&ltiframe&gt</code> through a Template Renderer website referenced in
+        the document. This arrangement allows for interactive selective disclosure
+        using OpenAttestationMerkleProofSignature2018.
+      </p>
+      <p>
+        The Template Renderer is a web application embedded in an iframe. It renders
+        verifiable credentials based on selected templates and must listen for
```
Comment (TallTed) [inline index.html:366]:
```suggestion
        verifiable credentials based on selected templates and needs to listen for
```

File: index.html [hunk 2]
```diff
@@ -350,6 +350,141 @@ <h4>SvgRenderingTemplate</h4>
...
+        specific messages from the Host application to facilitate the rendering
+        process.
+      </p>
+      <p>
+        There currently exists a number of OpenAttestationEmbeddedRenderer issuers
```
Comment (TallTed) [inline index.html:371]:
```suggestion
        There currently exist a number of OpenAttestationEmbeddedRenderer issuer
```

File: index.html [hunk 3]
```diff
@@ -350,6 +350,141 @@ <h4>SvgRenderingTemplate</h4>
...
+        and decentralized renderer implementations.
+      </p>
+      <p>
+        When an <a>issuer</a> desires to specify an embedded rendering instructions
```
Comment (TallTed) [inline index.html:375]:
```suggestion
        When an <a>issuer</a> desires to specify embedded rendering instructions
```

File: index.html [hunk 4]
```diff
@@ -395,6 +530,207 @@ <h5>Render (SvgRenderingTemplate)</h5>
...
         </ol>
       </section>
+
+      <section>
+        <h5>Action API (OpenAttestationEmbeddedRenderer)</h5>
+      
+        <p>
+          The Host is the application that displays the document with the help of the
+          Template Renderer. The Template Renderer MUST be a web application embedded
+          within an iframe specified by renderMethod.id. It MUST communicate with the
```
Comment (TallTed) [inline index.html:540]:
```suggestion
          within an `iframe` specified by `renderMethod`.`id`. It MUST communicate with the
```

File: index.html [hunk 5]
```diff
@@ -395,6 +530,207 @@ <h5>Render (SvgRenderingTemplate)</h5>
...
+          Host application using postMessage API to perform actions.
```
Comment (TallTed) [inline index.html:541]:
```suggestion
          Host application using `postMessage` API to perform actions.
```

File: index.html [hunk 6]
```diff
@@ -395,6 +530,207 @@ <h5>Render (SvgRenderingTemplate)</h5>
...
+                <th>action</th>
+              </tr>
+            </thead>
+            <tbody>
+              <tr>
+                <td>`GET_TEMPLATES`</td>
+                <td><pre>{ type: "GET_TEMPLATES" }</pre></td>
+                <td>
+                  Obtain a list of templates supported by the renderer for the given
+                  document. The list of templates is returned from UPDATE_TEMPLATES
```
Comment (TallTed) [inline index.html:603]:
```suggestion
                  document. The list of templates is returned to `UPDATE_TEMPLATES`
```

File: index.html [hunk 7]
```diff
@@ -395,6 +530,207 @@ <h5>Render (SvgRenderingTemplate)</h5>
...
+                  <pre>
+      {
+        type: "SELECT_TEMPLATE",
+        payload: "CUSTOM_TEMPLATE"
+      }
+                  </pre>
+                </td>
+                <td>
+                  Select the template to be used for rendering. It should be from the
+                  list returned by GET_TEMPLATES. If not found, a default template is
```
Comment (TallTed) [inline index.html:619]:
```suggestion
                  list returned by `GET_TEMPLATES`. If not found, a default template is
```

File: index.html [hunk 8]
```diff
@@ -395,6 +530,207 @@ <h5>Render (SvgRenderingTemplate)</h5>
...
+                  Render the verifiable credential inside the IFRAME using the
+                  selected template. Document is JSON object form of the verifiable
+                  credential.
+                </td>
+              </tr>
+              <tr>
+                <td>`PRINT`</td>
+                <td><pre>{ type: "PRINT" }</pre></td>
+                <td>
+                  MUST show the print dialog so the contents of the IFRAME can be
```
Comment (TallTed) [inline index.html:648]:
```suggestion
                  MUST show the print dialog so the contents of the iframe can be
```

File: index.html [hunk 9]
```diff
@@ -395,6 +530,207 @@ <h5>Render (SvgRenderingTemplate)</h5>
...
+            id: "template2",
+            label: "template2",
+          }
+        ]
+      }              
+                  </pre>
+                </td>
+                <td>
+                  Notify the Host of the list of template names that are usable from
+                  RENDER_METHOD, or GET_TEMPLATES calls.
```
Comment (TallTed) [inline index.html:727]:
```suggestion
                  `RENDER_METHOD` or `GET_TEMPLATES` calls.
```

**Review by TallTed — CHANGES_REQUESTED** [review]

**Review by cxcheng — APPROVED** [review]
We have discussed and are ok.

**Comment by cxcheng** [comment]
Updated with feedback from @msporny by moving the OpenAttestation Action API under the OpenAttestationEmbedderRenderer section.

**Comment by cxcheng** [comment]
> A group of us discussed this on the VC API call today.
> 
> General feedback was: the PR seems to be good enough to merge into the specification. The API section should probably go into the OAER section as a subsection. After that reformat is done we can merge this PR.

Change requested by @msporny done

**Review by msporny — APPROVED** [review]

**Pull request merged**